### PR TITLE
[compression] limit (de)compression, to network application limit size

### PR DESCRIPTION
**Pull request by bchocho**
### Description

Adding limits to compression:
* On the receiving end, a limit is needed to prevent malicious node from sending messages that decompress into a large size. Given lz4, decompression could have been as large as 2 GiB (i32::MAX).
* On the sending end, a limit is needed to ensure delivery of the compressed message to honest nodes.

Each call will have to set a limit; for network messages `MAX_APPLICATION_MESSAGE_SIZE` should be a good value as anyway the message needs to be sent over the network.

### Test Plan

Add unit test.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4556)

### CODE REVIEW

File: crates/aptos-compression/src/lib.rs [hunk 1]
```diff
@@ -108,6 +129,35 @@ pub fn decompress(
     Ok(raw_data)
 }
 
+// Derived from lz4 library
```
Comment (zekun000) [inline crates/aptos-compression/src/lib.rs:132]: we probably should attach a link here, also does compression mode changes how this is calculated? we should probably have upstream exposing the function public

**Review by zekun000 — APPROVED** [review]

File: crates/aptos-compression/src/lib.rs [hunk 2]
```diff
@@ -39,6 +40,7 @@ pub struct CompressionError(String);
 pub fn compress(
     raw_data: Vec<u8>,
     client: CompressionClient,
+    max_bytes: usize,
```
Comment (JoshLind) [inline crates/aptos-compression/src/lib.rs:43]: Given that we're only using compression for network messages, I'd just have this hard coded to `MAX_APPLICATION_MESSAGE_SIZE` and not dirty the call sites.

File: crates/aptos-compression/src/lib.rs [hunk 3]
```diff
@@ -78,27 +88,38 @@ pub fn compress(
 pub fn decompress(
     compressed_data: &CompressedData,
     client: CompressionClient,
+    max_size: usize,
```
Comment (JoshLind) [inline crates/aptos-compression/src/lib.rs:91]: Same here.

File: crates/aptos-compression/src/lib.rs [hunk 4]
```diff
@@ -78,27 +88,38 @@ pub fn compress(
...
-        Ok(raw_data) => raw_data,
+    // Check size of the data and initialize raw_data
+    let size = match get_decompressed_size(compressed_data, max_size) {
+        Ok(size) => size,
         Err(error) => {
             increment_compression_error(DECOMPRESS, client);
             return Err(CompressionError(format!(
-                "Failed to decompress the data: {:?}",
-                error.to_string()
+                "Failed to decompress the data: {}",
```
Comment (JoshLind) [inline crates/aptos-compression/src/lib.rs:102]: Can you fix the error message?

File: crates/aptos-compression/src/lib.rs [hunk 5]
```diff
@@ -50,12 +52,20 @@ pub fn compress(
...
-                error.to_string()
+                "Failed to compress the data: {}",
+                error
             )));
         }
     };
 
+    if compressed_data.len() > max_bytes {
+        return Err(CompressionError(format!(
+            "Compressed size too big: {} > {}",
```
Comment (JoshLind) [inline crates/aptos-compression/src/lib.rs:63]: nit: maybe `Compressed size greater than max. Size: {}, Max: {}`?

File: crates/aptos-compression/src/lib.rs [hunk 6]
```diff
@@ -108,6 +129,36 @@ pub fn decompress(
...
+        return Err(Error::new(
+            ErrorKind::InvalidInput,
+            "Source buffer must at least contain size prefix.",
+        ));
+    }
+
+    let size =
+        (src[0] as i32) | (src[1] as i32) << 8 | (src[2] as i32) << 16 | (src[3] as i32) << 24;
+
+    if size < 0 {
```
Comment (JoshLind) [inline crates/aptos-compression/src/lib.rs:145]: Do we think it makes sense to ever receive a message with `size == 0`?

File: crates/aptos-compression/src/lib.rs [hunk 7]
```diff
@@ -108,6 +129,36 @@ pub fn decompress(
...
+        return Err(Error::new(
+            ErrorKind::InvalidInput,
+            "Parsed size prefix in buffer must not be negative.",
+        ));
+    }
+
+    if size as usize > max_size {
```
Comment (JoshLind) [inline crates/aptos-compression/src/lib.rs:152]: nit: avoid the conversion twice and just say:
```
let size = size as usize;
```

File: crates/aptos-compression/src/lib.rs [hunk 8]
```diff
@@ -108,6 +129,36 @@ pub fn decompress(
...
+/// see: https://github.com/10XGenomics/lz4-rs/blob/0abc0a52af1f6010f9a57640b1dc8eb8d2d697aa/src/block/mod.rs#L162
+fn get_decompressed_size(src: &CompressedData, max_size: usize) -> std::io::Result<usize> {
+    if src.len() < 4 {
+        return Err(Error::new(
+            ErrorKind::InvalidInput,
+            "Source buffer must at least contain size prefix.",
+        ));
+    }
+
+    let size =
```
Comment (JoshLind) [inline crates/aptos-compression/src/lib.rs:142]: I'm surprised the crate does it like this. I'd have assumed there was an easier way? e.g., converting to a usize from bytes 🤔

**Review by JoshLind — APPROVED** [review]
Thanks @bchocho!

**Comment by bchocho** [comment]
@JoshLind I didn't want to spend too much time on this, so landing. But happy to follow-up. Tldr, I'm not sure if we want to keep compression as an independent crate or fold it into networking.
* Hard-coding `MAX_APPLICATION_MESSAGE_SIZE` doesn't work right now because it would create a circular dependency between networking and compression.
* Somewhat related, I don't expect network messages of size 0, but if we used this as a generic compression library, I could see it happening.

**Pull request merged**